src/train_cmd.cpp

```diff
@@ -3178,19 +3179,25 @@ bool TrainController(Train *v, Vehicle *nomove, bool reverse)
 						/* In front of a red signal */
 						Trackdir i = FindFirstTrackdir(trackdirbits);
 
-						/* Don't handle stuck trains here. */
+						/* Don't handle trains waiting at path signals here. */
 						if (HasBit(v->flags, VRF_TRAIN_STUCK)) return false;
 
 						if (!HasSignalOnTrackdir(gp.new_tile, ReverseTrackdir(i))) {
 							v->cur_speed = 0;
 							v->subspeed = 0;
 							v->progress = 255; // make sure that every bit of acceleration will hit the signal again, so speed stays 0.
-							if (!_settings_game.pf.reverse_at_signals || ++v->wait_counter < _settings_game.pf.wait_oneway_signal * DAY_TICKS * 2) return false;
+							++v->wait_counter;
```

Comment:
This does change the behaviour of wait_counter. And it's in a different order to the one below... Assuming this is intentional, comment to explain?

Reply:
True, will have to rework that.

Reply:
Looking at this again, I remember verifying that wait_counter is not used in the case where the behaviour differs, so the change in order of operations doesn't matter. I'm adding detailed comments for this.